client/src/explorer_common/state_and_memoizing.js

```diff
@@ -71,6 +71,34 @@ function root_reducer(state = initial_root_state, action) {
       }
     }
 
+    case "toggle_all": {
+      const { root } = payload;
+      const shouldExpand = !payload.is_expanded;
+
+      let nodes = [];
+      let search_nodes = (source) => {
+        if (source.children.length === 0) {
+          return;
+        }
+        _.forEach(source.children, (node) => {
+          if (
+            (shouldExpand && !node.isExpanded) ||
+            (!shouldExpand && node.isExpanded)
+          ) {
+            nodes.push(node.id);
+          }
+          search_nodes(node);
+        });
+      };
+      search_nodes(root);
```

Comment:
I left it in for the sake of it being a refactor, but the "search" part, i.e. `(shouldExpand && !isExpanded) || (!shouldExpand && isExpanded) ? id : []`, is actually an implementation bug.

If you expand any nodes before you click "expand all" then the expanded nodes end up collapsed, and vice versa. Guessing you interpreted `userExpanded`/`userCollapsed` as the ids of nodes that the current action wants to update, but they're actually become the new state rather than merging/applying to it.

Any confusion there's understandable though, the alternative interpretation does a better job of explaining the apparent redundancy between the two arrays. Track both user expanded and user collapsed nodes was Alex's solution to the problem that nodes a) can be open OR closed by default and b) the default can vary between nodes.